Bugzilla/DuoClient.pm

```diff
@@ -0,0 +1,264 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+# This Source Code Form is "Incompatible With Secondary Licenses", as
+# defined by the Mozilla Public License, v. 2.0.
+#
+# Adapted from https://github.com/duosecurity/duo_universal_python/blob/main/duo_universal/client.py
+
+package Bugzilla::DuoClient;
+
+use 5.10.1;
+use Moo;
+
+use Bugzilla::Constants;
+use Bugzilla::Error;
+use Bugzilla::Logging;
+use Bugzilla::Util qw(generate_random_password mojo_user_agent);
+
+use Mojo::JWT;
+use Mojo::URL;
+use Try::Tiny;
+use Types::Standard -types;
+
+use constant JTI_LENGTH             => 36;
+use constant MINIMUM_STATE_LENGTH   => 22;
+use constant MAXIMUM_STATE_LENGTH   => 1024;
+use constant JWT_DEFAULT_EXPIRATION => 300;
+
+use constant ERR_USERNAME     => 'The Duo username was invalid.';
+use constant ERR_CODE         => 'The Duo authorization code was missing.';
+use constant ERR_HEALTH_CHECK => 'The Duo service health check failed.';
+use constant ERR_TOKEN_ERROR =>
+  'The Duo service had an error obtaining authorization token.';
+use constant ERR_STATE_LEN => 'The Duo state must be at least '
+  . MINIMUM_STATE_LENGTH
+  . ' characters long and no longer than '
+  . MAXIMUM_STATE_LENGTH
+  . ' characters.';
+
+use constant OAUTH_V1_HEALTH_CHECK_ENDPOINT => '/oauth/v1/health_check';
+use constant OAUTH_V1_AUTHORIZE_ENDPOINT    => '/oauth/v1/authorize';
+use constant OAUTH_V1_TOKEN_ENDPOINT        => '/oauth/v1/token';
+use constant CLIENT_ASSERT_TYPE =>
+  'urn:ietf:params:oauth:client-assertion-type:jwt-bearer';
+
+#########################
+#    Initialization     #
+#########################
+
+has uri           => (is => 'ro', required => 1, isa => Str);
+has client_id     => (is => 'ro', required => 1, isa => Str);
+has client_secret => (is => 'ro', required => 1, isa => Str);
+has redirect_uri => (
+  is      => 'ro',
+  default => Bugzilla->localconfig->urlbase . 'mfa/duo/callback'
+);
+
+#########################
+#    Private Methods    #
+#########################
+
+# Standard JWT args to include with all JWT tokens
+sub _create_jwt_args {
+  my ($self, $endpoint) = @_;
+
+  my $jwt_args = {
+    'iss' => $self->client_id,
+    'sub' => $self->client_id,
```

Comment:
Question: The `sub` claim usually refers to `subject`, ie the user. Is passing the client ID here correct?

Reply:
Originally I was just going by https://github.com/duosecurity/duo_universal_python/blob/main/duo_universal/client.py#L120-L129 and what the Duo folks did in their example Python modules.

Also according to the Duo docs at https://duo.com/docs/oauthapi#json-web-tokens, it states that the 'iss' value should be 'The Client ID (or Integration key) from the application's page in the Duo Admin Panel'. And that the 'sub' value is described as `This should match the above iss value'.